feat: add doc about register plugin

Akarshit opened the pull request:
This partially solved #15 since a lot of what's sent to `registerPlugin` moves around in the context.

This add the docs about how to use the `registerPlugin` function and explain the arguments the function takes.

Akarshit committed:
feat: add doc about register plugin

Signed-off-by: Akarshit Wal <akarshitwal@gmail.com>

brent-hoover commented on docs/register-api-plugins.md:
````diff
@@ -0,0 +1,169 @@
+At a high level, an API plugin package is one where the main export is an async function that accepts a `ReactionAPI` instance and calls `api.registerPlugin` before returning. In a Reaction API project, your plugin package will be imported and used like this:
+
+```js
+import registerSomePlugin from "some-plugin";
````
Why is this `some-plugin`? That seems confusing

Akarshit replied:
update to "your-plugin"

brent-hoover commented on docs/register-api-plugins.md:
```diff
@@ -0,0 +1,169 @@
+  await app.registerPlugin({
+    label: "Plugin Example",
+    name: "plugin-example",
+    version: pkg.version,
```
Either include where `pkg` comes from or just include version

Akarshit replied:
Removed

brent-hoover commented on docs/register-api-plugins.md:
````diff
@@ -0,0 +1,169 @@
+```
+
+- `accountByUserId`: An async function with signature `(context, userId)` that must return an account document for the given `userId`, or `null` if one cannot be found. This will be used to set `context.account` and `context.accountId`.
+- `getHasPermissionFunctionForUser`: An async function with signature `(context)`, which must return an async function that checks permissions for `context.user` and returns a boolean. The signature of the returned function must be `(permissions, shopId)`, where `permissions` is an array of strings and `shopId` is an optional permission filter. The function must return `true` only if the context  ...
````
Should have spaces here. Difficult to read

Akarshit replied:
Done

brent-hoover commented on docs/register-api-plugins.md:
```diff
@@ -0,0 +1,169 @@
+
+### contextAdditions
+
+A plugin can add properties to context using the `contextAdditions` option when calling `registerPlugin`. They are added before any `preStartup` or `startup` functions are run.
```
I think we should mention that we generally recommend against doing this since it's easy to overwrite this should another plugin add something called the same thing. And at minimum they should probably namespace like `myplugin:thingIAmAddding`

brent-hoover reviewed (changes requested):
A few minor comments but this is GREAT!

MohanNarayana commented on docs/register-api-plugins.md:
```diff
@@ -0,0 +1,169 @@
+### Register API Plugins
```
Can we merge with this one about [developing plugins](https://github.com/reactioncommerce/docs/pull/47/files#diff-88390d67a4d665d22b449837fafcbe1b20e6b42a6850af71b3d0b662f2a68bda)

brent-hoover replied:
Although there is some overlap I don't think these are the same doc. Should definitely cross link though.

brent-hoover replied:
Actually I take that back. I think there needs to be a separate doc about `context` (as was the original ticket) and the other content here should be merged with that doc.

Akarshit commented:
@MohanNarayana @zenweasel I think the Developing API Plugin should should link to this doc for the following reasons:
1. Developing API plugin covers a lot of stuff about the linking and developing. This doc covers only the options that can be used while doing a `registerPlugin` call.
2. Adding this to the Developing API Plugin doc will make it too long.

Akarshit committed:
fix: better names and formatting

Signed-off-by: Akarshit Wal <akarshitwal@gmail.com>

Akarshit committed:
fix: reformatting to existing doc

Signed-off-by: Akarshit Wal <akarshitwal@gmail.com>

Akarshit commented:
I have updated this doc according to the format suggested by Missy. 
I have added Mohan's stuff on Developing API plugins from developer-concepts-2 branch, but have cut out a lot of stuff how core-api plugin works because it wasn't relevant.
I will give this another pass to make sure there is nothing present that's not relevant. Would appreciate some eyes on this one.

Akarshit commented on docs/plugins-environment-variables.md:
```diff
@@ -51,8 +46,300 @@ The plugins included with the development platform are:
+1. Start the whole system in standard mode using `make` command in the Reaction Development Platform directory.
+2. Run `make dev-<project folder name>` for one or more services to restart them in development mode. For example, `make dev-reaction` if you want to make API changes.
+
+But what exactly is this doing? Hopefully you don’t need to concern yourself with the details, but for those who want to know or who run into troubles, here’s the breakdown:
```
Do we actually need this? I think no.

Akarshit replied:
@zenweasel I think the details of how the switch happens are not useful to folks who are developing plugins. These details are only useful internally to understand linking and therefore should be removed from this doc.

brent-hoover replied:
Unclear on this comment since this is your PR?

Akarshit commented on docs/plugins-environment-variables.md:
```diff
@@ -51,8 +46,300 @@ The plugins included with the development platform are:
+### Switching Reaction to Development Mode
+
+Now let’s say you want to make some changes to a service and be able to see those changes reflected in the running service. By default, any changes you make do not affect the running service because all services run in standard mode. It’s done this way because it is much faster to start a service in standard mode. To develop a service, you need to switch it to development mode.
+Development mode differs from standard mode in the following ways:
```
Do we need this? Looks too much info to me.

Akarshit replied:
This again explains how the switch happens to development mode and what goes into making that switch. This looks out of scope for this document and should not be present here. I vote for removing this information.

Akarshit committed:
fix: update name and add basics

Signed-off-by: Akarshit Wal <akarshitwal@gmail.com>

brent-hoover commented:
Closing to be reopened in a later phase
